File: frontend/src/model/ai/datasets/dataset.ts
import { Course } from '../../course';
import { BaseTask } from '../../task/baseTask';
import { Task } from '../../task/task';
import { TaskGroup } from '../../task/taskGroup';

export interface ClearmlDataset {
  id: string;
  name: string;
  basename: string;
  description: string;
  created: string;
  dataset_stats: {
    file_count: number;
    total_size: number;
  };
}

export type DatasetStatus = 'saving' | 'processing' | 'completed' | 'failed';
export interface DatasetMetadata {
  class_map: {
    [key in string]: string;
  };
  is_grayscale: boolean;
  classes: string[];
  dimension: {
    x: number;
    y: number;
  };
  patch_size: number;
  patch_magnification: number;
  task_ids: number[];
}

export interface Dataset {
  id: string;
  name: string;
  description?: string;
  dataset_type: DatasetType;
  status: DatasetStatus;
  created_at: string;
  creator_id: string;
  metadata?: DatasetMetadata;
  clearml_dataset?: {
    id: string;
    name: string;
    project: {
      id: string;
    };
    runtime: {
      orig_dataset_name: string;
      orig_dataset_id: string;
      version: string;
      ds_file_count: number;
      ds_link_count: number;
      ds_total_size: number;
      ds_total_size_compressed: number;
      ds_change_add: number;
      ds_change_remove: number;
      ds_cange_modify: number;
      ds_change_size: number;
      status: string;
    };
  };
}

export type DatasetType = 'classification' | 'detection' | 'segmentation';

export type PatchSize = 128 | 256 | 512;
export type PatchMagnification = 0.2 | 0.5 | 1.0;

export interface CreateDataset {
  name: string;
  description?: string;
  type: DatasetType;
  file?: File;
}

export interface CreateOwnDataset {
  name: string;
  description?: string;
  type: DatasetType;
  patchSize: PatchSize;
  patchMagnification: PatchMagnification;
  tasks: {
    course: Course;
    taskGroup: TaskGroup;
    baseTask: BaseTask;
    task: Task;
  }[];
}

export type ConditionalDatasetMap<T> = {
  [key in DatasetType]: T[];
};

File: ai-api/app/schema/dataset.py
import datetime
from typing import Dict, Literal, Optional, List
from bson import ObjectId
from pydantic import BaseModel, Field
from app.schema.py_object_id import PyObjectId
from app.schema.base_mongo_model import BaseMongoModel

DatasetType = Literal["classification", "detection", "segmentation"]
DatasetStatus = Literal["saving", "processing", "completed", "failed"]


class DatasetDimension(BaseModel):
    x: int
    y: int


class DatasetMetadata(BaseModel):
    class_map: Optional[Dict]
    classes: Optional[List]
    is_grayscale: Optional[bool]
    dimension: Optional[DatasetDimension]


class Dataset(BaseMongoModel):
    id: PyObjectId = Field(default_factory=PyObjectId, alias="id")
    creator_id: str = Field(...)
    name: str = Field(...)
    description: Optional[str] = None
    dataset_type: DatasetType = Field(default="Classification")
    created_at: datetime.datetime = Field(...)
    status: Optional[DatasetStatus]
    clearml_dataset: Optional[Dict]
    metadata: Optional[DatasetMetadata]

    class Config:
        allow_population_by_field_name = True
        arbitrary_types_allowed = True
        json_encoders = {ObjectId: str}
        schema_extra = {
            "example": {
                "id": "9a201441-0d82-4d65-bd38-b2f14b700803",
                "name": "Test",
                "description": "That is a test project",
                "dataset_type": "Classification",
                "created_at": "2023-06-08T19:15:10.029+00:00",
                "status": "saving",
            }
        }
